pkg/hub/managedclusterset/controller.go

```diff
@@ -0,0 +1,265 @@
+func NewManagedClusterSetController(
+	clusterClient clientset.Interface,
+	clusterInformer clusterinformerv1.ManagedClusterInformer,
+	clusterSetInformer clusterinformerv1alpha1.ManagedClusterSetInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterSetController{
+		clusterClient:    clusterClient,
+		clusterLister:    clusterInformer.Lister(),
+		clusterSetLister: clusterSetInformer.Lister(),
+		eventRecorder:    recorder.WithComponentSuffix("managed-cluster-set-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterSetInformer.Informer()).
+		WithInformers(clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("ManagedClusterSetController", recorder)
+}
+
+func (c *managedClusterSetController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+
+	// if the sync is triggered by change of ManagedCluster, reconcile all cluster sets
+	if key == "key" {
+		if err := c.syncAllClusterSets(ctx); err != nil {
+			return fmt.Errorf("failed to sync all ManagedClusterSets: %w", err)
+		}
+		return nil
+	}
+
+	clusterSet, err := c.clusterSetLister.Get(key)
+	if errors.IsNotFound(err) {
+		// cluster set not found, could have been deleted, do nothing.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if err = c.syncClusterSet(ctx, clusterSet); err != nil {
+		return fmt.Errorf("failed to sync ManagedClusterSet %q: %w", key, err)
+	}
+
+	return nil
+}
+
+// syncAllClusterSets syncs all cluster sets
+func (c *managedClusterSetController) syncAllClusterSets(ctx context.Context) error {
+	clusterSets, err := c.clusterSetLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	errs := []error{}
+	for _, clusterSet := range clusterSets {
+		if err = c.syncClusterSet(ctx, clusterSet); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return operatorhelpers.NewMultiLineAggregate(errs)
+}
+
+// syncClusterSet syncs a particular cluster set
+func (c *managedClusterSetController) syncClusterSet(ctx context.Context, clusterSet *clusterv1alpha1.ManagedClusterSet) error {
+	klog.V(4).Infof("Reconciling ManagedClusterSet %s", clusterSet.Name)
+	// no work to do if the cluster set is deleted
+	if !clusterSet.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	selectedClusters := []string{}
+	selectedCondition := clusterv1alpha1.StatusCondition{
+		Type: clusterv1alpha1.ManagedClusterSetConditionEmpty,
+	}
+
+	selectors := clusterSet.Spec.ClusterSelectors
+	if len(selectors) == 0 {
+		selectors = append(selectors, clusterv1alpha1.ClusterSelector{})
+	}
+
+	for _, selector := range selectors {
+		clusters, err := c.selectClusters(selector)
+		if err != nil {
+			return err
+		}
+
+		selectedClusters = append(selectedClusters, clusters...)
+	}
+
+	selectedClusters = unique(selectedClusters)
```

Comment:
I don't see this actually doing anything.  Your plan is to add that later?  That has some impact on whether or not you fail open or closed if a particular expression doesn't work.

Reply:
@deads2k I think you might want to define "fail open" and "fail closed" here.

Reply:
Updated the code to fail closed.